sklearn/ensemble/tests/test_forest.py

```diff
@@ -373,6 +378,136 @@ def test_unfitted_feature_importances(name):
         getattr(FOREST_ESTIMATORS[name](), 'feature_importances_')
 
 
+@pytest.mark.parametrize("name", FOREST_CLASSIFIERS_REGRESSORS)
+def test_oob_importances(name):
+    # Check that oob permutation importances correctly identify that
+    # there are 3 important features
+
+    ForestEstimator = FOREST_ESTIMATORS[name]
+
+    clf = ForestEstimator(
+        n_estimators=10,
+        random_state=0,
+        feature_importances="permutation_oob",
+        bootstrap=True,
+    )
+    clf.fit(X_large, y_large)
+    importances = clf.feature_importances_
+    if name in FOREST_CLASSIFIERS:
+        imp_level = 0.025
+    else:
+        imp_level = 0.1
+
+    n_important = np.sum(importances > imp_level)
+    assert importances.shape[0] == 10
+    assert n_important == 3
+    assert np.all(importances[:3] > imp_level)
+
+
+@pytest.mark.parametrize("name", FOREST_CLASSIFIERS_REGRESSORS)
+def test_oob_importances_multi_class(name):
+    # Check that oob permutation importances correctly identify that
+    # there are 3 important features in a multi-class setting
+
+    ForestEstimator = FOREST_ESTIMATORS[name]
+
+    clf = ForestEstimator(
+        n_estimators=10,
+        random_state=0,
+        feature_importances="permutation_oob",
+        bootstrap=True,
+    )
+    clf.fit(X_large_multiclass, y_large_multiclass)
+    importances = clf.feature_importances_
+    if name in FOREST_CLASSIFIERS:
+        imp_level = 0.025
+    else:
+        imp_level = 0.1
+
+    n_important = np.sum(importances > imp_level)
+    assert importances.shape[0] == 10
+    assert n_important == 3
+    assert np.all(importances[:3] > imp_level)
+
+
+@pytest.mark.parametrize("name", FOREST_CLASSIFIERS_REGRESSORS)
+def test_oob_importance_ignores_random(name):
+    # Testing that a random feature with high cardinality registers as
+    # important using impurity-based feature importance but not out-of-bag
+    # permutation importance
+
+    X = X_large.copy()
```

Comment:
Let's separate regression/classification and we can generate the dataset in the function directly.